erpnext/accounts/doctype/budget/budget.py

```diff
@@ -51,56 +65,124 @@ def validate_accounts(self):
-				b.action_if_annual_budget_exceeded, b.action_if_accumulated_monthly_budget_exceeded
-			from `tabBudget` b, `tabBudget Account` ba
-			where
-				b.name=ba.parent and b.fiscal_year=%s and ba.account=%s and b.docstatus=1
-				and exists(select name from `tabCost Center` where lft<=%s and rgt>=%s and name=b.cost_center)
-		""", (args.fiscal_year, args.account, cc_lft, cc_rgt), as_dict=True)
-
-		for budget in budget_records:
-			if budget.budget_amount:
-				yearly_action = budget.action_if_annual_budget_exceeded
-				monthly_action = budget.action_if_accumulated_monthly_budget_exceeded
-
-				if monthly_action in ["Stop", "Warn"]:
-					budget_amount = get_accumulated_monthly_budget(budget.monthly_distribution,
-						args.posting_date, args.fiscal_year, budget.budget_amount)
-
-					args["month_end_date"] = get_last_day(args.posting_date)
-						
-					compare_expense_with_budget(args, budget.cost_center,
-						budget_amount, _("Accumulated Monthly"), monthly_action)
-
-				if yearly_action in ("Stop", "Warn") and monthly_action != "Stop" \
-					and yearly_action != monthly_action:
-						compare_expense_with_budget(args, budget.cost_center,
-							flt(budget.budget_amount), _("Annual"), yearly_action)
-
-def compare_expense_with_budget(args, cost_center, budget_amount, action_for, action):
-	actual_expense = get_actual_expense(args, cost_center)
 	if actual_expense > budget_amount:
 		diff = actual_expense - budget_amount
-		currency = frappe.db.get_value('Company', frappe.db.get_value('Cost Center',
-			cost_center, 'company'), 'default_currency')
+		currency = frappe.db.get_value('Company', args.company, 'default_currency')
 
-		msg = _("{0} Budget for Account {1} against Cost Center {2} is {3}. It will exceed by {4}").format(_(action_for),
-			frappe.bold(args.account), frappe.bold(cost_center),
+		msg = _("{0} Budget for Account {1} against {2} {3} is {4}. It will exceed by {5}").format(_(action_for),
+			frappe.bold(args.account), budget_against, frappe.bold(budget_against_type),
 			frappe.bold(fmt_money(budget_amount, currency=currency)), frappe.bold(fmt_money(diff, currency=currency)))
 
 		if action=="Stop":
 			frappe.throw(msg, BudgetError)
 		else:
 			frappe.msgprint(msg, indicator='orange')
 
+
+def get_actual_expense_for_CC(args, cost_center):
+	lft_rgt = frappe.db.get_value("Cost Center", cost_center, ["lft", "rgt"], as_dict=1)
+	args.update(lft_rgt)
+
+	condition = " and gle.posting_date <= %(month_end_date)s" if args.get("month_end_date") else ""
+
+	return flt(frappe.db.sql("""
+		select sum(gle.debit) - sum(gle.credit)
+		from `tabGL Entry` gle
+		where gle.account=%(account)s
+			and exists(select name from `tabCost Center`
+				where lft>=%(lft)s and rgt<=%(rgt)s and name=gle.cost_center)
+			and gle.fiscal_year=%(fiscal_year)s
+			and gle.company=%(company)s
+			and gle.docstatus=1
+			{condition}
+	""".format(condition=condition), (args))[0][0])
+
+
+def get_actual_expense_for_project(args):
```

Comment:
Again duplicate function, can be done within existing function passing appropriate condition